src/bindings/python/flux/job.py

```diff
@@ -85,3 +96,385 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+            time = duration
+        else:
+            raise ValueError("duration must be a float or string")
+        if time < 0:
+            raise ValueError("duration must not be negative")
+        if math.isnan(time) or math.isinf(time):
+            raise ValueError("duration must be a normal, finite value")
+        self.jobspec["attributes"]["system"]["duration"] = time
+
+    @property
+    def cwd(self):
+        """
+        Get working directory of job.
+        """
+        try:
+            self.jobspec["attributes"]["system"]["cwd"]
+        except KeyError:
+            return
+
+    @cwd.setter
+    def cwd(self, cwd):
+        """
+        Set working directory of job.
+        """
+        if not isinstance(cwd, six.string_types):
+            raise ValueError("cwd must be a string")
+        self.jobspec["attributes"]["system"]["cwd"] = cwd
+
+    @property
+    def environment(self):
+        """
+        Get (entire) environment of job.
+        """
+        try:
+            self.jobspec["attributes"]["system"]["environment"]
+        except KeyError:
+            return
+
+    @environment.setter
+    def environment(self, environ):
+        """
+        Set (entire) environment of job.
+        """
+        if not isinstance(environ, collectionsAbc.Mapping):
+            raise ValueError("environment must be a mapping")
+        self.jobspec["attributes"]["system"]["environment"] = environ
+
+    def _set_treedict(self, in_dict, key, val):
+        """
+        _set_treedict(d, "a.b.c", 42) is like d[a][b][c] = 42
+        but levels are created on demand.
+        """
+        path = key.split(".", 1)
+        if len(path) == 2:
+            self._set_treedict(in_dict.setdefault(path[0], {}), path[1], val)
+        else:
+            in_dict[key] = val
+
+    def setattr(self, key, val):
+        """
+        set job attribute
+        """
+        self._set_treedict(self.jobspec, "attributes." + key, val)
+
+    def setattr_shopt(self, key, val):
+        """
+        set job attribute: shell option
+        """
+        self.setattr("system.shell.options." + key, val)
+
+    def dumps(self):
+        return json.dumps(self.jobspec)
+
+    @property
+    def resources(self):
+        return self.jobspec.get("resources", None)
+
+    @property
+    def tasks(self):
+        return self.jobspec.get("tasks", None)
+
+    @property
+    def attributes(self):
+        return self.jobspec.get("attributes", None)
+
+    @property
+    def version(self):
+        return self.jobspec.get("version", None)
+
+    def resource_walk(self):
+        """
+        Traverse the resources in the `resources` section of the jobspec.
+
+        Performs a depth-first, pre-order traversal.
+        """
+
+        def walk_helper(res_list):
+            for resource in res_list:
+                yield resource
+                children = resource.get("with", [])
+                # TODO: convert to `yield from` after dropping 2.7
```

Comment:
*sigh* should we have issues and/or tags or a special string for this so we can do it wholesale?

Reply:
That makes a lot of sense. Do think we should do a multi-line string with the full py3 code that should replace it or just a common tag to search for like `# PY2: ....`

Reply:
Common search tag would be sufficient, a fixit like that would be nice but probably wont always be quite this small.